### ISSUE TITLE
Let's fix deep_munge <3 <3 <3

### ISSUE THREAD

**Issue by chancancode** [rails/rails#13420]
As a response to a security issue ([CVE-2012-2660](https://groups.google.com/forum/#!searchin/rubyonrails-security/deep_munge/rubyonrails-security/8SA-M3as7A8/Mr9fi9X4kNgJ), [CVE-2012-2694](https://groups.google.com/forum/#!searchin/rubyonrails-security/deep_munge/rubyonrails-security/jILZ34tAHF4/7x0hLH-o0-IJ) and [CVE-2013-0155](https://groups.google.com/forum/#!searchin/rubyonrails-security/CVE-2012-2660/rubyonrails-security/c7jT-EeN9eI/L0u4e87zYGMJ)), `deep_munge` [was introduced](https://github.com/imanel/rails/commit/060c91cd59ab86583a8f2f52142960d3433f62f5) as a solution to keep Rails secure by default.

Some of you feel strongly about this and are not satisfied with the solution. The good news is, #13188 has been merged to provide a temporary escape valve for those who know what they are doing. This will be available in Rails 4.1, so there's yet another reason to upgrade for those of you who are affected by this! (I encourage you to go through the advisories and make sure you *really* understand the issue before you fiddle with this.)

However, many of you are probably still dissatisfied with the current state of affairs. Personally, I admire your desire to push the framework forward and make it better for everyone.

Unfortunately, most of the solutions proposed so far do not meet the criteria @jeremy laid out in #13157. I'll reproduce the comment here:

```
Thanks for filing the PR, but for this to move forward, we need a much
stronger answer to the security vulnerability than just education.

We choose to be secure by default, and relying on education about
proper API use is begging for accidents.

So, this will not fly as-is. Please do keep pushing on this. We'd love
to have a compelling alternative that's secure by default.

* * *

To reiterate, fixes are welcome, so long as they address the security
vulnerability. Those are the constraints at play. Choosing to see that
as a plan to keep params parsing broken intentionally mischaracterizes
the issue and ensures zero progress on it. I encourage you to go deeper
than that! :grin:

For example, we could use strong_params to make a stronger statement
about the params we expect. We could declare what request content-types
our actions accept so the params parser would reject JSON requests
to actions that expect query string or www-urlencoded params. Etc.
```

In summary, "security by default" is a key cornerstone of Rails as a framework, and the core team is willing to aggressively defend this goal. Over the years, we have made some great progress in this area with mass assignment white-listing, strong parameters, automatic html escaping in erb, CSRF tokens, etc. I believe there is a similar opportunity here for us to push the envelope again as a community.

#13188 is going to buy us some time to explore different solutions to the problem. I encourage all the great minds in the community to gather here and start proposing solutions that meets the requirements @jeremy laid out. #13215 seems like a promising start, and I encourage you to help review the proposal and express your concerns.

**Please don't simply respond with +1s and "me too"s.** Those comments are only going to create noise for everyone and slow down the progress. **We heard you loud and clear, now it's time to get to work!**

Thank you in advance for your contribution :yellow_heart: :blue_heart: :purple_heart: :heart: :green_heart:

**Comment by jeremy** [comment]
Discussed with @sgrif and @chancancode. A Rails 5 path to eliminate params munging is to drop Active Record support for `foo: [1, 2, nil]` => `foo IN (1, 2) OR foo IS NULL`. Instead, drop the special `nil` handling and include `NULL` in the `IN` list: `foo IN (1, 2, NULL)`. Then databases do what we expect and we needn't munge params. We lose a convenient feature, but a very niche feature that can be worked around (and deprecated) by explicitly requesting `where(foo: [1, 2]).or(where(foo: nil))`

**Comment by sgrif** [comment]
Opened https://github.com/rails/rails/pull/18790 so we can discuss that potential path forward there.

**Comment by christhekeele** [comment]
:tada:

**Comment by kaspth** [comment]
Fixed by 52cf1a71b393486435fab4386a8663b146608996 :trollface:

**Comment by maclover7** [comment]
Looks ready to be closed to me.
